syntax = "proto2";

package RemyBuffers;

message WhiskerTree {
  optional MemoryRange domain = 1;

  repeated WhiskerTree children = 2;

  optional Whisker leaf = 3;

  optional ConfigRange config = 4;

  optional OptimizationSettings optimizer = 5;
}

message MemoryRange {
  optional Memory lower = 11;
  optional Memory upper = 12;
}

message Memory {
  optional double rec_send_ewma = 21;
  optional double rec_rec_ewma = 22;
  optional double rtt_ratio = 23;
  optional double slow_rec_rec_ewma = 24;
  //optional double loss_rate = 25;
}

message Whisker {
  optional sint32 window_increment = 31;
  optional double window_multiple = 32;
  optional double intersend = 33;

  optional MemoryRange domain = 34;
}

message OptimizationSetting {
  optional double min_value = 41;
  optional double max_value = 42;

  optional double min_change = 43;
  optional double max_change = 44;

  optional double multiplier = 45;

  optional double default_value = 46;
}

message OptimizationSettings {
  optional OptimizationSetting window_increment = 51;
  optional OptimizationSetting window_multiple = 52;
  optional OptimizationSetting intersend = 53;
}

message Range {
  optional double low = 61;
  optional double high = 62;
}

message ConfigRange {
  optional Range link_packets_per_ms = 71;
  optional Range rtt = 72;
  optional Range num_senders = 73;
  optional Range link_limit = 74; //CHANGED - new

  optional double mean_off_duration = 81;
  optional double mean_on_duration = 82;

  optional Range drop_rate = 83; //CHANGED - new
}

message NetConfig {
  optional double mean_on_duration = 1;
  optional double mean_off_duration = 2;
  optional uint32 num_senders = 3;
  optional double link_ppt = 4;
  optional double delay = 5;
  optional uint32 link_limit = 6; //CHANGED - new
  optional double drop_rate = 7; //CHANGED - new
}
